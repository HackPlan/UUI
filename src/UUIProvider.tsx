import React from 'react';

import { UUIComponentNames } from './components/UUIComponentNames';
import type { AccordionProps } from './components/Accordion/Accordion';
import type { AccordionPaneProps } from './components/Accordion/AccordionPane';
import type { BreadcrumbProps } from './components/Breadcrumb/Breadcrumb';
import type { ButtonProps } from './components/Button/Button';
import type { CascaderProps } from './components/Cascader/Cascader';
import type { CheckboxProps } from './components/Checkbox/Checkbox';
import type { CollapseProps } from './components/Collapse/Collapse';
import type { DatePickerProps } from './components/DatePicker/DatePicker';
import type { DialogProps } from './components/Dialog/Dialog';
import type { DrawerProps } from './components/Drawer/Drawer';
import type { IconProps } from './components/Icon/Icon';
import type { TextFieldProps } from './components/Input/TextField';
import type { NumberFieldProps } from './components/Input/NumberField';
import type { TextAreaProps } from './components/Input/TextArea';
import type { CountdownLabelProps } from './components/Label/CountdownLabel';
import type { DateLabelProps } from './components/Label/DateLabel';
import type { MoneyLabelProps } from './components/Label/MoneyLabel';
import type { NumberAbbrLabelProps } from './components/Label/NumberAbbrLabel';
import type { TimeLabelProps } from './components/Label/TimeLabel';
import type { LayoutProps } from './components/Layout/Layout';
import type { LayoutAsideProps } from './components/Layout/LayoutAside';
import type { LayoutFooterProps } from './components/Layout/LayoutFooter';
import type { LayoutHeaderProps } from './components/Layout/LayoutHeader';
import type { LayoutMainProps } from './components/Layout/LayoutMain';
import type { LayoutNavProps } from './components/Layout/LayoutNav';
import type { ListBoxProps } from './components/ListBox/ListBox';
import type { LoadingCoverProps } from './components/Loading/LoadingCover';
import type { LoadingSpinnerProps } from './components/Loading/LoadingSpinner';
import type { PageProps } from './components/Page/Page';
import type { PageAnnotatedSectionProps } from './components/Page/PageAnnotatedSection';
import type { PageSectionProps } from './components/Page/PageSection';
import type { PaginationProps } from './components/Pagination/Pagination';
import type { PageInfoProps } from './components/Pagination/PageInfo';
import type { PageJumperProps } from './components/Pagination/PageJumper';
import type { PageListProps } from './components/Pagination/PageList';
import type { PageNextButtonProps } from './components/Pagination/PageNextButton';
import type { PagePrevButtonProps } from './components/Pagination/PagePrevButton';
import type { PageSelectorProps } from './components/Pagination/PageSelector';
import type { PageSizeProps } from './components/Pagination/PageSize';
import type { PopoverProps } from './components/Popover/Popover';
import type { ProgressBarProps } from './components/ProgressBar/ProgressBar';
import type { RadioProps } from './components/Radio/Radio';
import type { RadioGroupProps } from './components/Radio/RadioGroup';
import type { SegmentControlProps } from './components/SegmentControl/SegmentControl';
import type { HTMLSelectProps } from './components/Select/HTMLSelect';
import type { SelectProps } from './components/Select/Select';
import type { SkeletonProps } from './components/Skeleton/Skeleton';
import type { SkeletonParagraphProps } from './components/Skeleton/Paragraph';
import type { SkeletonPictureProps } from './components/Skeleton/Picture';
import type { SkeletonTitleProps } from './components/Skeleton/Title';
import type { SliderProps } from './components/Slider/Slider';
import type { StepperProps } from './components/Stepper/Stepper';
import type { SwitchProps } from './components/Switch/Switch';
import type { TableProps } from './components/Table/Table';
import type { TabsProps } from './components/Tabs/Tabs';
import type { TabProps } from './components/Tabs/Tab';
import type { TagProps } from './components/Tag/Tag';
import type { ToastProps } from './components/Toast/Toast';
import type { ToasterProps } from './components/Toast/Toaster';


export interface UUIProviderCustomize {
  [UUIComponentNames.Accordion]?: AccordionProps['customize'];
  [UUIComponentNames.AccordionPane]?: AccordionPaneProps['customize'];
  [UUIComponentNames.Breadcrumb]?: BreadcrumbProps['customize'];
  [UUIComponentNames.Button]?: ButtonProps['customize'];
  [UUIComponentNames.Cascader]?: CascaderProps['customize'];
  [UUIComponentNames.Checkbox]?: CheckboxProps['customize'];
  [UUIComponentNames.Collapse]?: CollapseProps['customize'];
  [UUIComponentNames.DatePicker]?: DatePickerProps['customize'];
  [UUIComponentNames.Dialog]?: DialogProps['customize'];
  [UUIComponentNames.Drawer]?: DrawerProps['customize'];
  [UUIComponentNames.Icon]?: IconProps['customize'];
  [UUIComponentNames.TextField]?: TextFieldProps['customize'];
  [UUIComponentNames.NumberField]?: NumberFieldProps['customize'];
  [UUIComponentNames.TextArea]?: TextAreaProps['customize'];
  [UUIComponentNames.CountdownLabel]?: CountdownLabelProps['customize'];
  [UUIComponentNames.DateLabel]?: DateLabelProps['customize'];
  [UUIComponentNames.MoneyLabel]?: MoneyLabelProps['customize'];
  [UUIComponentNames.NumberAbbrLabel]?: NumberAbbrLabelProps['customize'];
  [UUIComponentNames.TimeLabel]?: TimeLabelProps['customize'];
  [UUIComponentNames.Layout]?: LayoutProps['customize'];
  [UUIComponentNames.LayoutAside]?: LayoutAsideProps['customize'];
  [UUIComponentNames.LayoutFooter]?: LayoutFooterProps['customize'];
  [UUIComponentNames.LayoutHeader]?: LayoutHeaderProps['customize'];
  [UUIComponentNames.LayoutMain]?: LayoutMainProps['customize'];
  [UUIComponentNames.LayoutNav]?: LayoutNavProps['customize'];
  [UUIComponentNames.ListBox]?: ListBoxProps['customize'];
  [UUIComponentNames.LoadingCover]?: LoadingCoverProps['customize'];
  [UUIComponentNames.LoadingSpinner]?: LoadingSpinnerProps['customize'];
  [UUIComponentNames.Page]?: PageProps['customize'];
  [UUIComponentNames.PageAnnotatedSection]?: PageAnnotatedSectionProps['customize'];
  [UUIComponentNames.PageSection]?: PageSectionProps['customize'];
  [UUIComponentNames.Pagination]?: PaginationProps['customize'];
  [UUIComponentNames.PageInfo]?: PageInfoProps['customize'];
  [UUIComponentNames.PageJumper]?: PageJumperProps['customize'];
  [UUIComponentNames.PageList]?: PageListProps['customize'];
  [UUIComponentNames.PageNextButton]?: PageNextButtonProps['customize'];
  [UUIComponentNames.PagePrevButton]?: PagePrevButtonProps['customize'];
  [UUIComponentNames.PageSelector]?: PageSelectorProps['customize'];
  [UUIComponentNames.PageSize]?: PageSizeProps['customize'];
  [UUIComponentNames.Popover]?: PopoverProps['customize'];
  [UUIComponentNames.ProgressBar]?: ProgressBarProps['customize'];
  [UUIComponentNames.Radio]?: RadioProps['customize'];
  [UUIComponentNames.RadioGroup]?: RadioGroupProps['customize'];
  [UUIComponentNames.SegmentControl]?: SegmentControlProps['customize'];
  [UUIComponentNames.HTMLSelect]?: HTMLSelectProps['customize'];
  [UUIComponentNames.Select]?: SelectProps['customize'];
  [UUIComponentNames.Skeleton]?: SkeletonProps['customize'];
  [UUIComponentNames.SkeletonParagraph]?: SkeletonParagraphProps['customize'];
  [UUIComponentNames.SkeletonPicture]?: SkeletonPictureProps['customize'];
  [UUIComponentNames.SkeletonTitle]?: SkeletonTitleProps['customize'];
  [UUIComponentNames.Slider]?: SliderProps['customize'];
  [UUIComponentNames.Stepper]?: StepperProps['customize'];
  [UUIComponentNames.Switch]?: SwitchProps['customize'];
  [UUIComponentNames.Table]?: TableProps['customize'];
  [UUIComponentNames.Tabs]?: TabsProps['customize'];
  [UUIComponentNames.Tab]?: TabProps['customize'];
  [UUIComponentNames.Tag]?: TagProps['customize'];
  [UUIComponentNames.Toast]?: ToastProps['customize'];
  [UUIComponentNames.Toaster]?: ToasterProps['customize'];
}

export type UUIProviderExtraCustomizeT = { [key: string]: any }

export interface UUIProviderContextValue<C extends UUIProviderExtraCustomizeT> {
  customize?: C & UUIProviderCustomize;
}
export const UUIProviderContext = React.createContext<UUIProviderContextValue<any> | null>(null)

export interface UUIProviderProps<C extends UUIProviderExtraCustomizeT> extends UUIProviderContextValue<C> {
  children: React.ReactNode;
}
export function UUIProvider<C extends UUIProviderExtraCustomizeT>(props: UUIProviderProps<C>) {
  return (
    <UUIProviderContext.Provider value={{
      customize: props.customize
    }}>
      {props.children}
    </UUIProviderContext.Provider>
  );
}
