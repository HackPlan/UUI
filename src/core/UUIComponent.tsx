import { IntrinsicNodeT, FunctionComponentNodeT, ClassComponentNodeT, IntrinsicNode, ComponentNode, UUIComponentNodes } from "./modules/UUICustomizeNode";
import React, { useMemo } from "react";
import { clone, isString, pickBy, mapKeys, isEmpty, mapValues } from "lodash-es";
import classNames from "classnames";
import { UUIComponentCustomizeProps, UUIConvenienceProps, UUIMetaProps } from "./modules/UUIComponentProps";
import { compileProps } from "./utils/compileProps";

/**
 * UUI Advanced Component for Function Component
 * @param options setup options
 * @param WrappedComponent wrapped function component
 */
export function UUIFunctionComponent<
  /**
   * Generic type P for target component props
   */
  P,
  /**
   * Generic type N for component node name
   */
  N extends string,
  /**
   * Generic type T for options.nodes value
   */
  T extends keyof IntrinsicNodeT | FunctionComponentNodeT | ClassComponentNodeT,
  /**
   * Generic type X for WrappedComponent generated nodes
   */
  X extends { [key in N]: T },
  /**
   * Generic type Z for component props.customize
   */
  Z extends UUIComponentCustomizeProps<X>
>(
  options: {
    prefix?: string;
    name: string;
    separator?: string;
    nodes: X;
  },
  WrappedComponent: (props: P, nodes: UUIComponentNodes<X>) => React.ReactElement,
) {
  const component: React.FunctionComponent<P & UUIConvenienceProps & UUIMetaProps & Z> = (props) => {
    const { prefix, separator } = props;
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const { nodes } = useMemo(() => {
      const finalOptions = getFinalOptions(options, { prefix, separator })
      const nodes = compileNodes(finalOptions)
      return { nodes }
    }, [prefix, separator])
    const compiledProps = compileProps(props)
    injectCustomizeProps(nodes, compiledProps);
    return WrappedComponent(compiledProps, nodes)
  }
  component.displayName = `<UUI> [Component] ${options.name}`
  return component
}

/**
 * UUI Advanced Component for Class Component
 * @param options setup options
 */
export function UUIClassComponent<
  /**
   * Generic type N for component node name
   */
  N extends string,
  /**
   * Generic type T for options.nodes value
   */
  T extends keyof IntrinsicNodeT | FunctionComponentNodeT | ClassComponentNodeT,
  /**
   * Generic type X for WrappedComponent generated nodes
   */
  X extends { [key in N]: T },
  /**
   * Generic type Z for component props.customize
   */
  Z extends UUIComponentCustomizeProps<X>
>(
  options: {
    prefix?: string;
    name: string;
    separator?: string;
    nodes: X;
  },
) {
  return class WrappedComponent<P = {}, S = {}, SS = any> extends React.Component<P & UUIConvenienceProps & UUIMetaProps & Z, S & { nodes: UUIComponentNodes<X> }, SS> {
    static displayName = `<UUI> [Component] ${options.name}`
    state: S & { nodes: UUIComponentNodes<X> }

    componentDidUpdate(prevProps: P & UUIConvenienceProps & UUIMetaProps & Z) {
      if (
        prevProps.prefix !== this.props.prefix ||
        prevProps.separator !== this.props.separator
      ) {
        const finalOptions = getFinalOptions(options, this.props)
        this.setState({ nodes: compileNodes(finalOptions) })
        const compiledProps = compileProps(this.props, (this.props as any).innerRef || undefined)
        injectCustomizeProps(this.state.nodes, compiledProps)
      }
    }

    constructor(props: P & UUIConvenienceProps & Z) {
      super(props)
      const finalOptions = getFinalOptions(options, props)
      this.state = { nodes: compileNodes(finalOptions) } as any
      this.state.nodes = compileNodes(finalOptions)
      const compiledProps = compileProps(props, (props as any).innerRef || undefined)
      injectCustomizeProps(this.state.nodes, compiledProps)
    }
  }
}

function getFinalOptions(options: any, props: any) {
  return {
    nodes: options.nodes,
    name: options.name,
    prefix: props.prefix || options.prefix || 'UUI',
    separator: props.separator || options.separator || '-',
  }
}

function compileNodes(options: any): any {
  return mapValues(options.nodes, (nodeElement, nodeName) => {
    if (isString(nodeElement)) {
      return IntrinsicNode(nodeElement as any, nodeName, options)
    } else {
      return ComponentNode(nodeElement as any, nodeName, options)
    }
  })
}

function injectCustomizeProps(nodes: any, props: any) {
  for (const nodeName of Object.keys(nodes)) {
    const customizeProps = props.customize && (props.customize as any)[nodeName]
    nodes[nodeName]['CustomizeProps'] = { customize: customizeProps }
  }
}
