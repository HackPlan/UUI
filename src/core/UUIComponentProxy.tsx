import React from 'react';
import { mergeCustomize } from './utils/mergeCustomize';
import { compileProps } from './utils/compileProps';

export function UUIComponentProxy<P extends { customize?: any }>(
  Component: React.ComponentType<P>,
  customize: P['customize'],
) {
  return (props: P) => {
    const compiledProps = compileProps(props)
    let finalCustomize: any = customize
    if (compiledProps.customize !== undefined) {
      finalCustomize = mergeCustomize(customize, compiledProps.customize)
    }
    return <Component {...props} customize={finalCustomize} />
  }
}
