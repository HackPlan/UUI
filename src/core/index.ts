export { UUIFunctionComponent, UUIClassComponent } from './UUIComponent';
export { UUIComponentProxy } from './UUIComponentProxy';

export type { UUIComponentProps, UUIFunctionComponentProps, UUIClassComponentProps } from './modules/UUIComponentProps';