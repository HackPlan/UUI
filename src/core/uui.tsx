/**
 * UUI Core Component Util
 *
 * this file provides multiple utils for building powerful uui components.
 * Any component built using this tool supports custom styles,
 * which means you can modify className, style and even children of nodes in the component.
 */


import React, { JSXElementConstructor, RefAttributes } from 'react';
import { mapValues, pick, isString, omit, merge, clone, uniq, isEmpty } from 'lodash';
import classNames from 'classnames';
import { mergeRefs } from '../utils/mergeRefs';

// ---------------------------------------------------------------
// Customize Extra Props
// ---------------------------------------------------------------

export interface NodeCustomizeClassNameProps {
  className?: string;
  overrideClassName?: string;
  extendClassName?: string;
}
export interface NodeCustomizeStyleProps {
  style?: React.CSSProperties;
  overrideStyle?: React.CSSProperties;
  extendStyle?: React.CSSProperties;
}
export interface NodeCustomizeChildrenProps {
  children?: React.ReactNode;
  overrideChildren?: React.ReactNode;
  extendChildrenBefore?: React.ReactNode;
  extendChildrenAfter?: React.ReactNode;
}
export type NodeCustomizeProps =
  & NodeCustomizeClassNameProps
  & NodeCustomizeStyleProps
  & NodeCustomizeChildrenProps
  & RefAttributes<any>

// ---------------------------------------------------------------
// Customize Extra Props Helper
// ---------------------------------------------------------------

function getCompiledNodeName(nodeName: string, options: { prefix: string; separator: string; name: string }) {
  return [options.prefix, options.name, nodeName].join(options.separator)
}
function getCompiledClassNames(nodeClassName: string, props: NodeCustomizeClassNameProps): string {
  return props.overrideClassName ? classNames(props.overrideClassName) : classNames(nodeClassName, props.className, props.extendClassName)
}
function getCompiledStyles(props: NodeCustomizeStyleProps): React.CSSProperties {
  return props.overrideStyle ? merge(props.overrideStyle) : merge(props.style, props.extendStyle)
}
function getCompiledChildren(props: NodeCustomizeChildrenProps): JSX.Element | null {
  if (props.overrideChildren) {
    return <>{props.overrideChildren}</>
  } else {
    return <>{props.extendChildrenBefore}{props.children}{props.extendChildrenAfter}</>
  }
}

// ---------------------------------------------------------------
// IntrinsicNode
// ---------------------------------------------------------------

type IntrinsicNodeCustomizeProps = NodeCustomizeProps
type IntrinsicNodeCustomizeOptions = {
  name: string;
  prefix: string;
  separator: string;
}

export type IntrinsicNodeT = JSX.IntrinsicElements
type IntrinsicNode<T extends keyof JSX.IntrinsicElements, N extends string | number | symbol> = (tagName: T, nodeName: N, options: IntrinsicNodeCustomizeOptions) => (props: JSX.IntrinsicElements[T]) => JSX.Element
function IntrinsicNode<T extends keyof JSX.IntrinsicElements, N extends string>(tagName: T, nodeName: N, options: IntrinsicNodeCustomizeOptions) {
  const Node = React.forwardRef((_props: JSX.IntrinsicElements[T], _ref) => {
    const customizeProps = (Node as any)['CustomizeProps'] as { customize?: IntrinsicNodeCustomizeProps } & UUIConvenienceProps
    const className = (() => {
      return getCompiledClassNames(getCompiledNodeName(nodeName, pick(options, 'name', 'prefix', 'separator')), {
        ...pick(customizeProps.customize, ['overrideClassName', 'extendClassName']),
        ...pick(_props, ['className']),
      })
    })()
    const style = (() => {
      const data = getCompiledStyles({
        ...pick(customizeProps.customize, ['overrideStyle', 'extendStyle']),
        ...pick(_props, ['style']),
      })
      return isEmpty(data) ? undefined : data
    })()
    const children = (() => {
      const noChildren = ['input', 'textarea', 'hr'].indexOf(tagName) !== -1
      const isSelectOption = tagName === 'option'
      // input tag do not support to pass children props
      let children: string | React.ReactNode | undefined = undefined
      if (isSelectOption) {
        // select option tag only support string type children,
        // if pass Fragments to children, it will show [Object Object] in html.
        children = _props.children
      } else if (!noChildren) {
        children = getCompiledChildren({
          ...pick(customizeProps.customize, ['overrideChildren', 'extendChildrenBefore', 'extendChildrenAfter'] as const),
          ...pick(_props, ['children']),
        })
      }
      return children
    })()

    /**
     * Merge both customize ref and component inner ref.
     */
    const ref = (() => {
      const refs: any[] = []
      const customizeRef = customizeProps.customize && customizeProps.customize.ref
      if (customizeRef) refs.push(customizeRef)
      if (_ref) refs.push(_ref)
      return mergeRefs(refs)
    })()

    /**
     * Merge both customize functions and component inner onXXX callback functions.
     */
    const mergedCallbackFunctions = (() => {
      const propsObj = _props as any
      const customizeObj = (customizeProps.customize || {}) as any

      const data: any = {}
      const attrs = uniq([
        ...Object.keys(propsObj),
        ...Object.keys(customizeObj),
      ])
      for (const attr of attrs) {
        if (attr.startsWith('on')) {
          const propsObjFunctionExist = !!(propsObj[attr] && typeof propsObj[attr] === 'function')
          const customizeObjFunctionExist = !!(customizeObj[attr] && typeof customizeObj[attr] === 'function')

          data[attr] = (...args: any[]) => {
            if (propsObjFunctionExist) {
              propsObj[attr](...args);
            }
            if (customizeObjFunctionExist) {
              customizeObj[attr](...args);
            }
          };
        }
      }
      return data
    })()

    return React.createElement(tagName, {
      ...omit(_props, 'children', 'ref', 'className', 'style'),
      ...mergedCallbackFunctions,
      ref,
      className, style,
    }, children)
  })
  Node.displayName = `<UUI> [IntrinsicNode] ${nodeName}`
  return Node
}

// ---------------------------------------------------------------
// ComponentNode
// ---------------------------------------------------------------

type ComponentNodeCustomizeProps<M extends string | number | symbol> = {
  [key in M]: IntrinsicNodeCustomizeProps
}
type ComponentNodeCustomizeOptions= {
  name: string;
  prefix: string;
  separator: string;
}

export type ComponentNodeT = (props: any, ...args: any) => any
type ComponentNode<P extends any, N extends string | number | symbol, M extends string | number | symbol> = (Target: React.ComponentType<P>, nodeName: N, options: ComponentNodeCustomizeOptions) => (props: P) => JSX.Element
function ComponentNode<P extends any, N extends string, M extends string>(Target: React.ComponentType<P>, nodeName: N, options: ComponentNodeCustomizeOptions) {
  const _Target = Target as any
  const Node = React.forwardRef((_props: P & ComponentNodeCustomizeProps<M>, ref) => {
    const customizeProps = (Node as any)['CustomizeProps'] as { customize?: ComponentNodeCustomizeProps<M> } & UUIConvenienceProps
    const nodeClassName = [options.prefix, options.name, nodeName].join(options.separator)

    return <_Target
      {...omit(_props, 'customize', 'ref')}
      ref={ref}
      className={classNames(nodeClassName, _props.className, customizeProps.className)}
      customize={merge(_props.customize, customizeProps.customize)}
    />
  })
  Node.displayName = `<UUI> [ComponentNode] ${nodeName}`
  return Node
}

// ---------------------------------------------------------------
// Type Helper
// ---------------------------------------------------------------

export type UUIComponentNodes<
  X extends { [key in string]?: keyof IntrinsicNodeT | ComponentNodeT },
> = {
  [key in keyof X]: X[key] extends keyof IntrinsicNodeT
    ? ReturnType<IntrinsicNode<X[key], key>>
    : (X[key] extends ComponentNodeT ? X[key] : never)
}
export type UUIComponentCustomizeProps<
  X extends { [key in string]?: keyof IntrinsicNodeT | ComponentNodeT },
> = {
  /**
   * Customize component nodes
   * @default none
   */
  customize?: {
    [key in keyof X]?: X[key] extends keyof IntrinsicNodeT
      ? NodeCustomizeProps & Partial<JSX.IntrinsicElements[X[key]]>
      : (X[key] extends ComponentNodeT ? NonNullable<Parameters<X[key]>[0]['customize']> : never)
  };
}
export type UUIConvenienceProps = {
  /**
   * Convenience className props,
   * this props will be applied to append to extendClassName of component Root node customize props.
   * @default none
   */
  className?: string;
  /**
   * Convenience style props,
   * this props will be applied to merge to extendStyle of component Root node customize props.
   * @default none
   */
  style?: React.CSSProperties;
}
export type UUIComponentProps<P, X extends { [key in string]?: keyof IntrinsicNodeT | ComponentNodeT }> = P & UUIConvenienceProps & UUIComponentCustomizeProps<X>
export type UUIFunctionComponentProps<T extends (...args: any) => any> = Parameters<T>[0]
export type UUIClassComponentProps<T extends JSXElementConstructor<any>> = React.ComponentProps<T>

// ---------------------------------------------------------------
// UUI Component Util
// ---------------------------------------------------------------

export abstract class UUI {
  /**
   * UUI Advanced Component for Function Component
   * @param options setup options
   * @param WrappedComponent wrapped function component
   */
  static FunctionComponent<
    /**
     * Generic type P for target component props
     */
    P,
    /**
     * Generic type N for component node name
     */
    N extends string,
    /**
     * Generic type T for options.nodes value
     */
    T extends keyof IntrinsicNodeT | ComponentNodeT,
    /**
     * Generic type X for WrappedComponent generated nodes
     */
    X extends { [key in N]: T },
    /**
     * Generic type Z for component props.customize
     */
    Z extends UUIComponentCustomizeProps<X>
  >(
    options: {
      prefix?: string;
      name: string;
      separator?: string;
      nodes: X;
    },
    WrappedComponent: (props: P, nodes: UUIComponentNodes<X>) => React.ReactElement,
  ) {
    const finalOptions: Required<typeof options> = getFinalOptions(options)
    const nodes = compileNodes(finalOptions)
    const component: React.FunctionComponent<P & UUIConvenienceProps & Z> = (props) => {
      const compiledProps = compileProps(props, finalOptions, undefined)
      injectCustomizeProps(nodes, compiledProps)
      return WrappedComponent(compiledProps, nodes)
    }
    component.displayName = `<UUI> [Component] ${finalOptions.name}`
    return component
  }

  /**
   * UUI Advanced Component for Class Component
   * @param options setup options
   */
  static ClassComponent<
    /**
     * Generic type N for component node name
     */
    N extends string,
    /**
     * Generic type T for options.nodes value
     */
    T extends keyof IntrinsicNodeT | ComponentNodeT,
    /**
     * Generic type X for WrappedComponent generated nodes
     */
    X extends { [key in N]: T },
    /**
     * Generic type Z for component props.customize
     */
    Z extends UUIComponentCustomizeProps<X>
  >(
    options: {
      prefix?: string;
      name: string;
      separator?: string;
      nodes: X;
    },
  ) {
    const finalOptions: Required<typeof options> = getFinalOptions(options)
    const nodes = compileNodes(finalOptions)
    return class WrappedComponent<P = {}, S = {}, SS = any> extends React.Component<P & UUIConvenienceProps & Z, S, SS> {
      nodes: UUIComponentNodes<X>
      displayName: string
      constructor(props: P & UUIConvenienceProps & Z) {
        super(props)
        this.displayName = `<UUI> [Component] ${finalOptions.name}`
        const compiledProps = compileProps(props, finalOptions, (props as any).innerRef || undefined)
        injectCustomizeProps(nodes, compiledProps)
        this.nodes = nodes
      }
    }
  }
}

function getFinalOptions(options: any) {
  return {
    ...options,
    prefix: options.prefix || 'UUI',
    separator: options.separator || '-',
  }
}

function compileProps(props: any, options: any, ref: any): any {
  /**
   * Convenience props: className, style
   * className will be injected into customize.Root { extendClassName: ... }
   * style will be injected into customize.Root { extendStyle: ... }
   */
  const compiledProps = clone(props)
  if (!compiledProps.customize) {
    compiledProps.customize = {}
  }

  // Generally, UUI component should contain a Root node.
  if (
    (options.nodes as any)['Root'] &&
    isString((options.nodes as any)['Root'])
  ) {
    const rootCustomizeProps: any = (compiledProps.customize as any)['Root'] || {};
    if (compiledProps.className) rootCustomizeProps.extendClassName = classNames(compiledProps.className, rootCustomizeProps.extendClassName);
    if (compiledProps.style) rootCustomizeProps.extendStyle = Object.assign(compiledProps.style, rootCustomizeProps.extendStyle) as any;
    (compiledProps.customize as any)['Root'] = rootCustomizeProps;
  }
  compiledProps.ref = ref

  return  compiledProps
}

function compileNodes(options: any): any {
  return mapValues(options.nodes, (nodeElement, nodeName) => {
    if (isString(nodeElement)) {
      return IntrinsicNode(nodeElement as any, nodeName, options)
    } else {
      return ComponentNode(nodeElement as any, nodeName, options)
    }
  })
}

function injectCustomizeProps(nodes: any, props: any) {
  for (const nodeName of Object.keys(nodes)) {
    const customizeProps = props.customize && (props.customize as any)[nodeName]
    nodes[nodeName]['CustomizeProps'] = { customize: customizeProps }
  }
}