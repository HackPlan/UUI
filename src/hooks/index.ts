export { default as useFocusTrap } from '@charlietango/use-focus-trap'
export { useLockBodyScroll } from 'react-use'