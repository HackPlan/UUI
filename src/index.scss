@import './styles/global.scss';

@import '../node_modules/normalize.css/normalize.css';
@import './components/Button/Button.scss';
@import './components/Checkbox/Checkbox.scss';
@import './components/Dialog/Dialog.scss';
@import './components/Drawer/Drawer.scss';
@import './components/Icon/Icon.scss';
@import './components/Input/TextField.scss';
@import './components/Input/TextArea.scss';
@import './components/Input/NumberField.scss';
@import './components/Pagination/PageInfo.scss';
@import './components/Pagination/PageJumper.scss';
@import './components/Pagination/PageList.scss';
@import './components/Popover/Popover.scss';
@import './components/Radio/Radio.scss';
@import './components/Select/HTMLSelect.scss';
@import './components/Select/Select.scss';
@import './components/Skeleton/Skeleton.scss';
@import './components/Slider/Slider.scss';
@import './components/Stepper/Stepper.scss';
@import './components/Switch/Switch.scss';
@import './components/Table/Table.scss';
@import './components/Tag/Tag.scss';
@import './components/Toast/Toaster.scss';