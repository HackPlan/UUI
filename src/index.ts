import './styles/tailwind.css'
import './index.scss'

export * from './components/HelloWorld'
export * from './components/Switch'
export * from './components/Tag'
export * from './components/Skeleton'
export * from './components/Page'

