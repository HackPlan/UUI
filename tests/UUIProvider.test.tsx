import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { UUIFunctionComponent, UUIFunctionComponentProps } from '../src/core';
import { UUIProvider } from '../src/UUIProvider';

Enzyme.configure({ adapter: new Adapter })

const TestComponent = UUIFunctionComponent({
  name: 'TestComponent',
  nodes: {
    Root: 'div',
    Node: 'div',
    ExtendTest: 'div',
  }
}, (props: {}, nodes) => {
  const { Root, Node, ExtendTest } = nodes
  return (
    <Root className="root-base-classname" style={{
      backgroundColor: 'green',
      fontWeight: 300,
    }}>
      <Node className="node-base-classname" style={{
        width: 200,
        height: 300,
      }} />
      <ExtendTest>
        <div>Original</div>
      </ExtendTest>
    </Root>
  )
})
type TestComponentProps = UUIFunctionComponentProps<typeof TestComponent>

type TestUUIProviderCustomize = {
  TestComponent?: TestComponentProps['customize'];
}

const globalCustomizeConfig: TestUUIProviderCustomize = {
  TestComponent: {
    Root: {
      extendClassName: 'GlobalExtendButtonClassName',
      extendStyle: {
        backgroundColor: 'red',
        color: 'blue',
        fontWeight: 400,
      }
    },
  }
}

/**
 * UUI Provider
 *
 * 测试 UUIProvider 是否正确合并 customize 属性
 */
it('UUIProvider', () => {

  const tree1 = renderer
  .create(
    <UUIProvider
      customize={{
        Button: {
          Root: {
            extendClassName: ''
          }
        }
      }}
    >
      <TestComponent />
    </UUIProvider>
  )
  .toJSON();
  expect(tree1).toMatchSnapshot();

  const tree2 = renderer
  .create(
    <UUIProvider<TestUUIProviderCustomize>
      customize={globalCustomizeConfig}
    >
      <TestComponent />
    </UUIProvider>
  )
  .toJSON();
  expect(tree2).toMatchSnapshot();


  const tree3 = renderer
  .create(
    <UUIProvider<TestUUIProviderCustomize>
      customize={globalCustomizeConfig}
    >
      <TestComponent
        customize={{
          Root: {
            extendStyle: {
              color: 'yellow',
              fontSize: 34,
            },
          }
        }}
      />
      <TestComponent
        customize={{
          Root: {
            extendStyle: {
              color: 'yellow',
              fontSize: 34,
            },
          }
        }}
      />
    </UUIProvider>
  )
  .toJSON();
  expect(tree3).toMatchSnapshot();

})