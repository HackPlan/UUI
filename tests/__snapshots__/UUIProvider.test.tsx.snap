// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UUIProvider 1`] = `
<div
  className="UUI-TestComponent-Root root-base-classname"
  style={
    Object {
      "backgroundColor": "green",
      "fontWeight": 300,
    }
  }
>
  <div
    className="UUI-TestComponent-Node node-base-classname"
    style={
      Object {
        "height": 300,
        "width": 200,
      }
    }
  />
  <div
    className="UUI-TestComponent-ExtendTest"
  >
    <div>
      Original
    </div>
  </div>
</div>
`;

exports[`UUIProvider 2`] = `
<div
  className="UUI-TestComponent-Root root-base-classname GlobalExtendButtonClassName"
  style={
    Object {
      "backgroundColor": "red",
      "color": "blue",
      "fontWeight": 400,
    }
  }
>
  <div
    className="UUI-TestComponent-Node node-base-classname"
    style={
      Object {
        "height": 300,
        "width": 200,
      }
    }
  />
  <div
    className="UUI-TestComponent-ExtendTest"
  >
    <div>
      Original
    </div>
  </div>
</div>
`;

exports[`UUIProvider 3`] = `
Array [
  <div
    className="UUI-TestComponent-Root root-base-classname GlobalExtendButtonClassName"
    style={
      Object {
        "backgroundColor": "red",
        "color": "yellow",
        "fontSize": 34,
        "fontWeight": 400,
      }
    }
  >
    <div
      className="UUI-TestComponent-Node node-base-classname"
      style={
        Object {
          "height": 300,
          "width": 200,
        }
      }
    />
    <div
      className="UUI-TestComponent-ExtendTest"
    >
      <div>
        Original
      </div>
    </div>
  </div>,
  <div
    className="UUI-TestComponent-Root root-base-classname GlobalExtendButtonClassName"
    style={
      Object {
        "backgroundColor": "red",
        "color": "yellow",
        "fontSize": 34,
        "fontWeight": 400,
      }
    }
  >
    <div
      className="UUI-TestComponent-Node node-base-classname"
      style={
        Object {
          "height": 300,
          "width": 200,
        }
      }
    />
    <div
      className="UUI-TestComponent-ExtendTest"
    >
      <div>
        Original
      </div>
    </div>
  </div>,
]
`;
